File: ios/sdk/swift/internal/ContactBridge.swift
import CrossPL

extension ContactInternal {

/* @CrossClass */
open class ContactBridge: CrossBase {
  public static let TAG = "elastos"
  
  init() {
    super.init(className: String(describing: ContactBridge.self))
  }

  deinit {
    for(_, channel) in mCustomChannelMap {
      channel.unbind()
    }
    mCustomChannelMap.removeAll();
    mListener?.unbind()
    mDataListener?.unbind()
  }
  
  public func appendChannelStrategy(channelStrategy: Contact.ChannelStrategy) -> Int {
    let oldChannelStrategy = mCustomChannelMap[channelStrategy.getChannelId()]
    oldChannelStrategy?.unbind()
    
    channelStrategy.bind()
    let ret = appendChannelStrategy(channelId: channelStrategy.getChannelId(),
                                    channelStrategy: channelStrategy as CrossBase)
    if(ret < 0) {
        channelStrategy.unbind();
        return ret;
    }

    mCustomChannelMap[channelStrategy.getChannelId()] = channelStrategy
    return 0;

  }
  
  public func setListener(listener: Contact.Listener?) {
    if(mListener != nil) {
      mListener!.unbind()
    }
    mListener = listener

    mListener!.bind()
    setListener(listener: mListener as CrossBase?)
  }
  
  public func setDataListener(listener: Contact.DataListener?) {
    if(mDataListener != nil) {
      mDataListener!.unbind()
    }
    mDataListener = listener

    mDataListener!.bind()
    setDataListener(listener: mDataListener as CrossBase?)
  }

  public func setHumanInfo(humanCode: String, item: Contact.HumanInfo.Item, value: String) -> Int {
    return setHumanInfo(humanCode: humanCode, item: item.rawValue, value: value)
  }
  
  public func setUserInfo(item: Contact.UserInfo.Item, value: String) -> Int {
    let ret = setHumanInfo(humanCode: "-user-info-", item: item.rawValue, value: value);
    return ret
  }

  public func setIdentifyCode(type: Contact.UserInfo.Kind, value: String) -> Int {
    let ret = setIdentifyCode(type: type.rawValue, value: value);
    return ret;
  }

  public func getUserInfo() -> Contact.UserInfo? {
    if(mListener == nil) {
      return nil
    }
    
    var sbInfo = String()
    var ret = getHumanInfo(humanCode: "-user-info-", info: &sbInfo)
    if(ret < 0) {
      Log.w(tag: ContactBridge.TAG, msg: "Failed to get user info. ret=\(ret)")
      return nil
    }

    let userInfo = Contact.UserInfo();
    ret = userInfo.fromJson(info: sbInfo);
    if(ret < 0) {
      Log.w(tag: ContactBridge.TAG, msg: "Failed to deserialize user info. ret=\(ret)")
      return nil
    }

    return userInfo;
  }

  public func listFriendInfo() -> [Contact.FriendInfo]? {
    assert(mListener != nil)

    var json = String()
    let ret = getFriendList(info: &json);
    if(ret < 0) {
      Log.w(tag: ContactBridge.TAG, msg: "Failed to list friend info. ret=\(ret)")
      return nil
    }

//    TypeToken<List<Contact.FriendInfo.FriendJson>> friendInfoListType = new TypeToken<List<Contact.FriendInfo.FriendJson>>(){};
//    let listJson = new Gson().fromJson(json.toString(), friendInfoListType.getType());
    let listJson = try! JSONDecoder().decode([Contact.FriendInfo.FriendJson].self, from: json.data(using: .utf8)!)

    var list = [Contact.FriendInfo]()
    for it in listJson {
      let friendInfo = Contact.FriendInfo()
      let ret = friendInfo.fromJson(info: it)
      if(ret < 0) {
        Log.w(tag: ContactBridge.TAG, msg: "Failed to deserialize friend info. ret=\(ret)")
        return nil
      }

      list.append(friendInfo)
    }

    return list
  }

  public func listFriendCode() -> [String] {
    var friendCodeList = [String]()

    let friendList = listFriendInfo()
    if(friendList == nil) {
      return friendCodeList
    }

    for it in friendList! {
      friendCodeList.append(it.humanCode!)
    }

    return friendCodeList
  }

  public func getStatus(humanCode: String) -> Contact.Status? {
    let ret = getHumanStatus(humanCode: humanCode)
    if(ret < 0) {
      return nil
    }

    return Contact.Status(rawValue: ret)
  }

//  public func makeMessage(type: ContactMessage.Kind, data: Data, cryptoAlgorithm: String?) -> Contact.Message {
//    let msg = Contact.Message(type: type, data: data, cryptoAlgorithm: cryptoAlgorithm)
//    return msg
//  }

  public static func MakeTextMessage(text: String, cryptoAlgorithm: String?) -> Contact.Message {
    let msg = Contact.Message(text: text, cryptoAlgorithm: cryptoAlgorithm)
    return msg
  }

  public static func MakeFileMessage(file: URL, cryptoAlgorithm: String?) -> Contact.Message {
    let msg = Contact.Message(file: file,
                              cryptoAlgorithm: cryptoAlgorithm)
    return msg
  }

  public func sendMessage(friendCode: String, channelType: Contact.Channel, message: Contact.Message) -> Int {
    var ret = message.syncMessageToNative()
    if(ret < 0) {
      return ret;
    }

    ret = sendMessage(friendCode: friendCode, channelType: channelType.rawValue, message: message)

    return ret;
  }
  
  public func pullFileAsync(friendCode: String, channelType: Contact.Channel,
                            fileInfo: Contact.Message.FileData) -> Int {
    guard fileInfo.devId != nil else {
      return -1;
    }
    
    let ret = pullData(friendCode: friendCode, channelType: channelType.rawValue,
                       dev: fileInfo.devId!, dataId: fileInfo.toString())

    return ret;
  }

  public func cancelPullFile(friendCode: String, channelType: Contact.Channel,
                             fileInfo: Contact.Message.FileData) -> Int {
    guard fileInfo.devId != nil else {
      return -1;
    }
    
    let ret = cancelPullData(friendCode: friendCode, channelType: channelType.rawValue,
                             dev: fileInfo.devId!, dataId: fileInfo.toString())

    return ret;
  }

  /* @CrossNativeInterface */
  public static func IsDidFriend(friendCode: String) -> Bool {
    let ret = crosspl_Proxy_ContactBridge_IsDidFriend(friendCode)
    return ret
  }
  
  /* @CrossNativeInterface */
  public func start() -> Int {
    let ret = crosspl_Proxy_ContactBridge_start(nativeHandle)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func stop() -> Int {
    let ret = crosspl_Proxy_ContactBridge_stop(nativeHandle)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func addFriend(friendCode: String, summary: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_addFriend(nativeHandle, friendCode, summary)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func removeFriend(friendCode: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_removeFriend(nativeHandle, friendCode)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func acceptFriend(friendCode: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_acceptFriend(nativeHandle, friendCode)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func syncInfoDownloadFromDidChain() -> Int {
    let ret = crosspl_Proxy_ContactBridge_syncInfoDownloadFromDidChain(nativeHandle)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func syncInfoUploadToDidChain() -> Int {
    let ret = crosspl_Proxy_ContactBridge_syncInfoUploadToDidChain(nativeHandle)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func setWalletAddress(name: String, value: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_setWalletAddress(nativeHandle, name, value)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func appendChannelStrategy(channelId: Int, channelStrategy: CrossBase?) -> Int {
    let ret = crosspl_Proxy_ContactBridge_appendChannelStrategy(nativeHandle,
                                                                Int32(channelId),
                                                                channelStrategy)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func setListener(listener: CrossBase?) {
    crosspl_Proxy_ContactBridge_setListener(nativeHandle, listener)
  }
  
  /* @CrossNativeInterface */
  private func setDataListener(listener: CrossBase?) {
    crosspl_Proxy_ContactBridge_setDataListener(nativeHandle, listener)
  }
  
  /* @CrossNativeInterface */
  private func setHumanInfo(humanCode: String, item: Int, value: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_setHumanInfo(nativeHandle, humanCode, Int32(item), value)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func setIdentifyCode(type: Int, value: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_setIdentifyCode(nativeHandle, Int32(type), value)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public func getHumanInfo(humanCode: String, info: inout String) -> Int {
    var nsinfo = info as NSString?
    
    let ret = crosspl_Proxy_ContactBridge_getHumanInfo(nativeHandle, humanCode, &nsinfo)
    if ret < 0 {
      return Int(ret)
    }
    info = nsinfo! as String
    
    return 0
  }
  
  /* @CrossNativeInterface */
  public func getFriendList(info: inout String) -> Int {
    var nsinfo = info as NSString?
    let ret = crosspl_Proxy_ContactBridge_getFriendList(nativeHandle, &nsinfo)
    if ret < 0 {
      return Int(ret)
    }
    info = nsinfo! as String
    
    return 0
  }
  
  /* @CrossNativeInterface */
  public func getHumanStatus(humanCode: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_getHumanStatus(nativeHandle, humanCode)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func sendMessage(friendCode: String, channelType: Int, message: CrossBase) -> Int {
    let ret = crosspl_Proxy_ContactBridge_sendMessage(nativeHandle, friendCode, Int32(channelType), message)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func pullData(friendCode: String, channelType: Int, dev: String, dataId: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_pullData(nativeHandle, friendCode, Int32(channelType),
                                                   dev, dataId)
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  private func cancelPullData(friendCode: String, channelType: Int, dev: String, dataId: String) -> Int {
    let ret = crosspl_Proxy_ContactBridge_cancelPullData(nativeHandle, friendCode, Int32(channelType),
                                                         dev, dataId)
    return Int(ret)
  }
  
  private var mCustomChannelMap = [Int: Contact.ChannelStrategy]()
  private var mListener: Contact.Listener?
  private var mDataListener: Contact.DataListener?
}

}

File: android/sdk/src/main/java/org/elastos/sdk/elephantwallet/contact/internal/ContactBridge.java
package org.elastos.sdk.elephantwallet.contact.internal;

import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.elastos.sdk.elephantwallet.contact.Contact;
import org.elastos.sdk.elephantwallet.contact.Utils;
import org.elastos.tools.crosspl.CrossBase;
import org.elastos.tools.crosspl.annotation.CrossClass;
import org.elastos.tools.crosspl.annotation.CrossInterface;

import java.io.File;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

@CrossClass
class ContactBridge extends CrossBase {
    public class SyncInfoLocation {
        public static final int DidChain = 1;
        public static final int Oss = 2;
    }

    public static final String TAG = "elastos";

    protected ContactBridge() {
        super(ContactBridge.class.getName(), 0);
    }

    public void finalize() {
        for (Contact.ChannelStrategy channel : mCustomChannelMap.values()) {
            channel.unbind();
        }
        mCustomChannelMap.clear();

        if (mListener != null) {
            mListener.unbind();
        }
        if (mDataListener != null) {
            mDataListener.unbind();
        }

        super.finalize();
    }

    public synchronized int appendChannelStrategy(Contact.ChannelStrategy channelStrategy) {
        Contact.ChannelStrategy oldChannelStrategy = mCustomChannelMap.get(channelStrategy.getChannelId());
        if (oldChannelStrategy != null) {
            oldChannelStrategy.unbind();
        }

        channelStrategy.bind();
        int ret = appendChannelStrategy((CrossBase) channelStrategy);
        if (ret < 0) {
            channelStrategy.unbind();
            return ret;
        }

        mCustomChannelMap.put(channelStrategy.getChannelId(), channelStrategy);
        return 0;
    }

    public synchronized void setListener(Contact.Listener listener) {
        if (mListener != null) {
            mListener.unbind();
        }
        mListener = listener;

        mListener.bind();
        setListener((CrossBase) mListener);
    }

    public synchronized void setDataListener(Contact.DataListener listener) {
        if (mDataListener != null) {
            mDataListener.unbind();
        }
        mDataListener = listener;

        mDataListener.bind();
        setDataListener((CrossBase) mDataListener);
    }

    public int setIdentifyCode(IdentifyCode.Type type, String value) {
        int ret = setIdentifyCode(type.id(), value);
        return ret;
    }

    public int setUserInfo(HumanInfo.Item item, String value) {
        int ret = setHumanInfo("-user-info-", item.id(), value);
        return ret;
    }

    public Contact.UserInfo getUserInfo() {
        assert (mListener != null);

        Contact.UserInfo userInfo = new Contact.UserInfo();
        int ret = getHumanInfo("-user-info-", userInfo);
        if (ret < 0) {
            Log.w(TAG, "Failed to get user info. ret=" + ret);
            return null;
        }

        return userInfo;
    }

    public int getUserBrief(StringBuffer brief) {
        assert (mListener != null);

        int ret = getHumanBrief("-user-info-", UserInfo.GetCurrDevId(), brief);
        if (ret < 0) {
            Log.w(TAG, "Failed to get user info. ret=" + ret);
            return ret;
        }

        return 0;
    }

    public List<Contact.FriendInfo> listFriendInfo() {
        assert(mListener != null);

        StringBuffer json = new StringBuffer();
        int ret = getFriendList(json);
        if(ret < 0) {
            Log.w(TAG, "Failed to list friend info. ret=" + ret);
            return null;
        }

        TypeToken<List<Contact.FriendInfo.FriendJson>> friendInfoListType = new TypeToken<List<Contact.FriendInfo.FriendJson>>(){};
        List<Contact.FriendInfo.FriendJson> listJson = Utils.GetGsonBuilder().create().fromJson(json.toString(), friendInfoListType.getType());

        List<Contact.FriendInfo> list = new ArrayList<>();
        for(Contact.FriendInfo.FriendJson it: listJson) {
            Contact.FriendInfo friendInfo = new Contact.FriendInfo();
            ret  = friendInfo.fromJson(it);
            if(ret < 0) {
                Log.w(TAG, "Failed to deserialize friend info. ret=" + ret);
                return null;
            }

            list.add(friendInfo);
        }

        return list;
    }

    public List<String> listFriendCode() {
        List<String> friendCodeList = new ArrayList<String>();

        List<Contact.FriendInfo> friendList = listFriendInfo();
        if(friendList == null) {
            return friendCodeList;
        }

        for(Contact.FriendInfo it: friendList) {
            friendCodeList.add(it.humanCode);
        }

        return friendCodeList;
    }

    public int setHumanInfo(String humanCode, HumanInfo.Item item, String value) {
        int ret = setHumanInfo(humanCode, item.id(), value);
        return ret;
    }

    public <T extends HumanInfo> int getHumanInfo(String humanCode, T humanInfo) {
        assert(mListener != null);

        StringBuffer sbInfo = new StringBuffer();
        int ret = getHumanInfo(humanCode, sbInfo);
        if(ret < 0) {
            Log.w(TAG, "Failed to get human info. ret=" + ret);
            return ret;
        }

        ret  = humanInfo.fromJson(sbInfo.toString());
        if(ret < 0) {
            Log.w(TAG, "Failed to deserialize human info. ret=" + ret);
            return ret;
        }

        return 0;
    }

    public String findAvatarFile(String avatar) {
        StringBuffer filepath = new StringBuffer();
        int ret = findAvatarFile(avatar, filepath);
        if(ret < 0) {
            return null;
        }

        return filepath.toString();
    }

    public Contact.Status getStatus(String humanCode) {
        int ret = getHumanStatus(humanCode);
        if(ret < 0) {
            return null;
        }

        return Contact.Status.valueOf(ret);
    }

    public int sendMessage(String friendCode, ContactChannel channelType, Contact.Message message) {
        if(message == null) {
            return -1;
        }

        int ret = message.syncMessageToNative();
        if(ret < 0) {
            return ret;
        }

        ret = sendMessage(friendCode, channelType.id(), message);
        if(ret < 0) {
            return ret;
        }

        return ret;
    }

    public int pullFileAsync(String friendCode, ContactChannel channelType,
                        Contact.Message.FileData fileInfo) {
        if(fileInfo == null) {
            return -1;
        }

        int ret = pullData(friendCode, channelType.id(), fileInfo.devId, fileInfo.toString());

        return ret;
    }

    public int cancelPullFile(String friendCode, ContactChannel channelType,
                              Contact.Message.FileData fileInfo) {
        if(fileInfo == null) {
            return -1;
        }

        int ret = cancelPullData(friendCode, channelType.id(), fileInfo.devId, fileInfo.toString());

        return ret;
    }

    public static Contact.Message MakeTextMessage(String text, String cryptoAlgorithm, String memo) {
        Contact.Message msg = new Contact.Message(text, cryptoAlgorithm, memo);
        return msg;
    }

    public static Contact.Message MakeBinaryMessage(byte[] binary, String cryptoAlgorithm, String memo) {
        Contact.Message msg = new Contact.Message(binary, cryptoAlgorithm, memo);
        return msg;
    }

    public static Contact.Message MakeFileMessage(File file, String cryptoAlgorithm, String memo) {
        Contact.Message msg = new Contact.Message(file, cryptoAlgorithm, memo);
        return msg;
    }


//    private static Contact.Message MakeMessage(ContactMessage.Type type, byte[] data, String cryptoAlgorithm) {
//        Contact.Message msg = new Contact.Message(type, data, cryptoAlgorithm);
//        return msg;
//    }

    @CrossInterface
    public static native boolean IsDidFriend(String friendCode);

    @CrossInterface
    public native int start();

    @CrossInterface
    public native int stop();

    @CrossInterface
    public native int addFriend(String friendCode, String summary);

    @CrossInterface
    public native int removeFriend(String friendCode);

    @CrossInterface
    public native int acceptFriend(String friendCode);

    @CrossInterface
    public native int syncInfoDownloadFromDidChain();

    @CrossInterface
    public native int syncInfoUploadToDidChain();

    @CrossInterface
    public native int syncInfoDownload(int fromLocation);

    @CrossInterface
    public native int syncInfoUpload(int toLocation);

    @CrossInterface
    public native int exportUserData(String toFile);

    @CrossInterface
    public native int importUserData(String fromFile);

    @CrossInterface
    public native int setWalletAddress(String name, String value);

    @CrossInterface
    private native int appendChannelStrategy(CrossBase channelStrategy);

    @CrossInterface
    private native void setListener(CrossBase listener);

    @CrossInterface
    private native void setDataListener(CrossBase listener);

    @CrossInterface
    private native int setIdentifyCode(int item, String value);

    @CrossInterface
    private native int setHumanInfo(String humanCode, int item, String value);

    @CrossInterface
    private native int getHumanInfo(String humanCode, StringBuffer info);

    @CrossInterface
    private native int getHumanBrief(String friendCode, String devId, StringBuffer brief);

    @CrossInterface
    public native int findAvatarFile(String avatar, StringBuffer filepath);

    @CrossInterface
    private native int getFriendList(StringBuffer info);

    @CrossInterface
    private native int getHumanStatus(String humanCode);

    @CrossInterface
    private native int sendMessage(String friendCode, int channelType, CrossBase message);

    @CrossInterface
    private native int pullData(String friendCode, int channelType, String devId, String dataId);

    @CrossInterface
    private native int cancelPullData(String friendCode, int channelType, String devId, String dataId);

    private HashMap<Integer, Contact.ChannelStrategy> mCustomChannelMap = new HashMap<>();
    private Contact.Listener mListener;
    private Contact.DataListener mDataListener;

    static { Utils.EnsureNativeLibrary(); }
}
